File: health_sync/providers/oura.py
from __future__ import annotations

import json
import secrets
from datetime import UTC, datetime, timedelta
from typing import Any
from urllib.parse import quote, urlencode, urlparse

import requests

from ..config import LoadedConfig, require_str
from ..db import HealthSyncDb
from ..util import (
    basic_auth_header,
    dt_to_iso_z,
    iso_to_dt,
    parse_yyyy_mm_dd,
    oauth_listen_for_code,
    open_in_browser,
    request_json,
    sha256_hex,
    utc_now_iso,
)


OURA_PROVIDER = "oura"
OURA_BASE = "https://api.ouraring.com"
OURA_OAUTH_AUTHORIZE_DEFAULT = "https://moi.ouraring.com/oauth/v2/ext/oauth-authorize"
OURA_OAUTH_TOKEN_DEFAULT = "https://moi.ouraring.com/oauth/v2/ext/oauth-token"


def _oura_default_redirect_uri() -> str:
    # Oura accepts `http://localhost/...` for local redirect URIs but rejects
    # `http://127.0.0.1/...` with `400 invalid_request`.
    return "http://localhost:8080/callback"


def _oura_scopes(cfg: LoadedConfig) -> str:
    # Keep it broad; users can override if their app is configured with different scopes.
    return cfg.config.oura.scopes


def _oura_authorize_base(cfg: LoadedConfig) -> str:
    return cfg.config.oura.authorize_url or OURA_OAUTH_AUTHORIZE_DEFAULT


def _oura_token_base(cfg: LoadedConfig) -> str:
    return cfg.config.oura.token_url or OURA_OAUTH_TOKEN_DEFAULT


def _oidc_discovery_urls(issuer: str) -> list[str]:
    iss = issuer.strip().rstrip("/")
    if not iss:
        return []

    parsed = urlparse(iss)
    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
        return []

    out: list[str] = [f"{iss}/.well-known/openid-configuration"]
    # Oura callbacks may include an anonymous issuer path segment that differs
    # from the token/authorize endpoint base by one path element.
    if parsed.path.endswith("/oauth-anonymous"):
        parent = iss[: -len("/oauth-anonymous")]
        out.append(f"{parent}/.well-known/openid-configuration")

    seen: set[str] = set()
    deduped: list[str] = []
    for u in out:
        if u in seen:
            continue
        seen.add(u)
        deduped.append(u)
    return deduped


def _oura_discover_endpoints(
    sess: requests.Session,
    *,
    issuer: str | None,
) -> tuple[str | None, str | None, str | None]:
    if not issuer:
        return None, None, None

    for discovery_url in _oidc_discovery_urls(issuer):
        try:
            discovery = request_json(sess, "GET", discovery_url, max_retries=1)
        except Exception:  # noqa: BLE001
            continue
        if not isinstance(discovery, dict):
            continue

        authorize = discovery.get("authorization_endpoint")
        token = discovery.get("token_endpoint")
        authorize_s = authorize.strip() if isinstance(authorize, str) and authorize.strip() else None
        token_s = token.strip() if isinstance(token, str) and token.strip() else None
        if token_s:
            return authorize_s, token_s, discovery_url

    return None, None, None


def _oura_resolve_token_endpoint(
    sess: requests.Session,
    cfg: LoadedConfig,
    *,
    issuer: str | None = None,
    token_extra: dict[str, Any] | None = None,
) -> tuple[str, str | None]:
    if token_extra:
        saved_endpoint = token_extra.get("token_endpoint")
        if isinstance(saved_endpoint, str) and saved_endpoint.strip():
            return saved_endpoint.strip(), None
        if issuer is None:
            saved_issuer = token_extra.get("issuer")
            if isinstance(saved_issuer, str) and saved_issuer.strip():
                issuer = saved_issuer.strip()

    _auth, discovered_token, discovery_url = _oura_discover_endpoints(sess, issuer=issuer)
    if discovered_token:
        return discovered_token, discovery_url

    return _oura_token_base(cfg), None


def _oura_redirect(cfg: LoadedConfig) -> tuple[str, int, str]:
    redirect_uri = cfg.config.oura.redirect_uri or _oura_default_redirect_uri()
    u = urlparse(redirect_uri)
    if u.scheme not in ("http", "https"):
        raise RuntimeError(f"Invalid `oura.redirect_uri`: {redirect_uri}")
    host = u.hostname or "127.0.0.1"
    port = u.port or (443 if u.scheme == "https" else 80)
    path = u.path or "/callback"
    return redirect_uri, port, path


def oura_auth(db: HealthSyncDb, cfg: LoadedConfig, *, listen_host: str = "127.0.0.1", listen_port: int = 0) -> None:
    client_id = require_str(cfg, cfg.config.oura.client_id, key="oura.client_id")
    client_secret = require_str(cfg, cfg.config.oura.client_secret, key="oura.client_secret")
    redirect_uri, redirect_port, callback_path = _oura_redirect(cfg)

    if listen_port == 0:
        listen_port = redirect_port

    state = secrets.token_urlsafe(16)
    scope = _oura_scopes(cfg)
    authorize_base = _oura_authorize_base(cfg)

    auth_url = f"{authorize_base}?{urlencode({'response_type': 'code', 'client_id': client_id, 'redirect_uri': redirect_uri, 'scope': scope, 'state': state}, quote_via=quote)}"

    print("Open this URL to authorize Oura:")
    print("Authorize URL uses `client_id` + exact `redirect_uri`; `client_secret` is only used in token exchange.")
    print(auth_url)
    open_in_browser(auth_url)

    res = oauth_listen_for_code(listen_host=listen_host, listen_port=listen_port, callback_path=callback_path)
    if res.error:
        raise RuntimeError(f"Oura auth error: {res.error}")
    if res.state and res.state != state:
        raise RuntimeError("Oura auth failed: state mismatch")

    sess = requests.Session()
    token_endpoint, discovery_url = _oura_resolve_token_endpoint(sess, cfg, issuer=res.issuer)
    if discovery_url:
        print(f"Using Oura OIDC discovery endpoint: {discovery_url}")
        print(f"Using Oura token endpoint: {token_endpoint}")

    try:
        token = request_json(
            sess,
            "POST",
            token_endpoint,
            headers={
                "Authorization": basic_auth_header(client_id, client_secret),
                "Content-Type": "application/x-www-form-urlencoded",
            },
            data={
                "grant_type": "authorization_code",
                "code": res.code,
                "redirect_uri": redirect_uri,
            },
        )
    except RuntimeError as e:
        msg = str(e)
        if "invalid_grant" in msg or "invalid_request" in msg:
            raise RuntimeError(
                f"{msg}. Oura auth codes are short-lived and single-use. Request a fresh code and retry once at {token_endpoint}."
            ) from e
        raise

    access_token = token["access_token"]
    refresh_token = token.get("refresh_token")
    token_type = token.get("token_type")
    scope_resp = token.get("scope")
    expires_in = token.get("expires_in")
    expires_at = None
    if isinstance(expires_in, int):
        expires_at = dt_to_iso_z(datetime.now(UTC) + timedelta(seconds=int(expires_in)))

    extra = {k: v for k, v in token.items() if k not in {"access_token", "refresh_token", "token_type", "scope", "expires_in"}}
    extra["token_endpoint"] = token_endpoint
    if res.issuer:
        extra["issuer"] = res.issuer
    if discovery_url:
        extra["oidc_discovery_url"] = discovery_url

    db.set_oauth_token(
        provider=OURA_PROVIDER,
        access_token=access_token,
        refresh_token=refresh_token,
        token_type=token_type,
        scope=scope_resp,
        expires_at=expires_at,
        extra=extra,
    )
    print("Stored Oura OAuth token in DB.")


def _oura_refresh_if_needed(db: HealthSyncDb, cfg: LoadedConfig, sess: requests.Session) -> str:
    tok = db.get_oauth_token(OURA_PROVIDER)
    if not tok:
        raise RuntimeError(
            "Missing Oura credentials. Run `health-sync auth oura` "
            f"(config: {cfg.path})."
        )

    access_token = tok["access_token"]
    refresh_token = tok.get("refresh_token")
    expires_at = tok.get("expires_at")
    token_extra = tok.get("extra") if isinstance(tok.get("extra"), dict) else {}

    if not refresh_token:
        raise RuntimeError(
            "Stored Oura token is missing `refresh_token`. Oura now requires OAuth2 tokens with refresh support; "
            "run `health-sync auth oura` again."
        )

    try:
        exp = iso_to_dt(expires_at) if expires_at else datetime.now(UTC) - timedelta(days=1)
    except Exception:  # noqa: BLE001
        exp = datetime.now(UTC) - timedelta(days=1)

    if exp - datetime.now(UTC) > timedelta(seconds=60):
        return access_token

    client_id = require_str(cfg, cfg.config.oura.client_id, key="oura.client_id")
    client_secret = require_str(cfg, cfg.config.oura.client_secret, key="oura.client_secret")
    token_endpoint, discovery_url = _oura_resolve_token_endpoint(sess, cfg, token_extra=token_extra)

    try:
        token = request_json(
            sess,
            "POST",
            token_endpoint,
            headers={
                "Authorization": basic_auth_header(client_id, client_secret),
                "Content-Type": "application/x-www-form-urlencoded",
            },
            data={
                "grant_type": "refresh_token",
                "refresh_token": refresh_token,
            },
        )
    except RuntimeError as e:
        msg = str(e)
        if "invalid_grant" in msg or "invalid_request" in msg:
            raise RuntimeError(
                f"{msg}. Oura refresh token appears invalid for current endpoint. Run `health-sync auth oura` to get a fresh code/token pair."
            ) from e
        raise

    new_access = token["access_token"]
    new_refresh = token.get("refresh_token") or refresh_token
    token_type = token.get("token_type")
    scope_resp = token.get("scope")
    expires_in = token.get("expires_in")
    new_expires_at = None
    if isinstance(expires_in, int):
        new_expires_at = dt_to_iso_z(datetime.now(UTC) + timedelta(seconds=int(expires_in)))

    merged_extra = {k: v for k, v in token_extra.items() if k not in {"access_token", "refresh_token", "token_type", "scope", "expires_in"}}
    merged_extra.update({k: v for k, v in token.items() if k not in {"access_token", "refresh_token", "token_type", "scope", "expires_in"}})
    merged_extra["token_endpoint"] = token_endpoint
    if discovery_url:
        merged_extra["oidc_discovery_url"] = discovery_url

    db.set_oauth_token(
        provider=OURA_PROVIDER,
        access_token=new_access,
        refresh_token=new_refresh,
        token_type=token_type,
        scope=scope_resp,
        expires_at=new_expires_at,
        extra=merged_extra,
    )
    return new_access


def _oura_headers(access_token: str) -> dict[str, str]:
    return {
        "Authorization": f"Bearer {access_token}",
        "User-Agent": "health-sync/0.1 (+local sqlite cache)",
        "Accept": "application/json",
    }


def _oura_fetch_all(
    sess: requests.Session,
    *,
    access_token: str,
    path: str,
    params: dict[str, str] | None = None,
) -> list[dict]:
    params = dict(params or {})
    url = f"{OURA_BASE}{path}"

    items: list[dict] = []
    next_token: str | None = None
    while True:
        if next_token:
            params["next_token"] = next_token

        j = request_json(sess, "GET", url, headers=_oura_headers(access_token), params=params)

        if "data" not in j:
            # e.g. personal_info
            if isinstance(j, dict):
                items.append(j)
            break

        data = j.get("data")
        if isinstance(data, list):
            items.extend([x for x in data if isinstance(x, dict)])
        elif isinstance(data, dict):
            items.append(data)

        next_token = j.get("next_token")
        if not next_token:
            break

    return items


def oura_sync(db: HealthSyncDb, cfg: LoadedConfig) -> None:
    sess = requests.Session()
    access_token = _oura_refresh_if_needed(db, cfg, sess)

    start_date = cfg.config.oura.start_date
    today_date = datetime.now(UTC).date()
    today = today_date.isoformat()
    # Oura's `/sleep` endpoint needs end_date one day ahead to reliably include
    # the latest completed night in the returned day buckets.
    sleep_end_date = (today_date + timedelta(days=1)).isoformat()
    overlap_days = int(cfg.config.oura.overlap_days)

    # Daily + session collections use date windows; Oura doesn't provide a general updated_since flag.
    resources: list[tuple[str, str]] = [
        ("daily_activity", "/v2/usercollection/daily_activity"),
        ("daily_sleep", "/v2/usercollection/daily_sleep"),
        ("daily_readiness", "/v2/usercollection/daily_readiness"),
        ("sleep", "/v2/usercollection/sleep"),
        ("workout", "/v2/usercollection/workout"),
    ]

    # Heartrate is time-series; we keep it, but only backfill up to the same start date.
    ts_resources: list[tuple[str, str]] = [
        ("heartrate", "/v2/usercollection/heartrate"),
    ]

    print("Syncing Oura...")

    # personal_info (single object)
    with db.sync_run(provider=OURA_PROVIDER, resource="personal_info") as run:
        with db.transaction():
            items = _oura_fetch_all(sess, access_token=access_token, path="/v2/usercollection/personal_info")
            if items:
                op = db.upsert_record(
                    provider=OURA_PROVIDER,
                    resource="personal_info",
                    record_id="me",
                    payload=items[0],
                    start_time=None,
                    end_time=None,
                    source_updated_at=None,
                    fetched_at=utc_now_iso(),
                )
                run.add_upsert(op)
            db.set_sync_state(provider=OURA_PROVIDER, resource="personal_info", watermark=utc_now_iso())

    # date-window collections
    for name, path in resources:
        with db.sync_run(provider=OURA_PROVIDER, resource=name) as run:
            with db.transaction():
                state = db.get_sync_state(provider=OURA_PROVIDER, resource=name)
                if state and state.watermark:
                    # Re-fetch a small overlap to catch backfilled/edited data.
                    wm_dt = iso_to_dt(state.watermark)
                    start_dt = wm_dt - timedelta(days=overlap_days)
                    start_s = start_dt.date().isoformat()
                else:
                    start_s = start_date

                end_s = sleep_end_date if name == "sleep" else today
                params = {"start_date": start_s, "end_date": end_s}
                items = _oura_fetch_all(sess, access_token=access_token, path=path, params=params)

                for item in items:
                    stable = json.dumps(item, sort_keys=True, separators=(",", ":"), ensure_ascii=True)
                    record_id = str(item.get("id") or item.get("day") or item.get("timestamp") or sha256_hex(stable))
                    start_time = item.get("day") or item.get("start_datetime") or item.get("timestamp") or item.get("start_time")
                    end_time = item.get("end_datetime") or item.get("end_time")
                    updated_at = item.get("updated_at") or item.get("modified_at") or item.get("timestamp")
                    op = db.upsert_record(
                        provider=OURA_PROVIDER,
                        resource=name,
                        record_id=record_id,
                        payload=item,
                        start_time=start_time,
                        end_time=end_time,
                        source_updated_at=updated_at,
                    )
                    run.add_upsert(op)

                # Canonical UTC watermark for this resource.
                db.set_sync_state(provider=OURA_PROVIDER, resource=name, watermark=utc_now_iso())

    # time-series collections (HR): delta by last stored timestamp.
    for name, path in ts_resources:
        with db.sync_run(provider=OURA_PROVIDER, resource=name) as run:
            with db.transaction():
                max_ts = db.get_max_start_time(provider=OURA_PROVIDER, resource=name)
                if max_ts:
                    try:
                        start_dt = iso_to_dt(max_ts) - timedelta(days=1)
                    except Exception:  # noqa: BLE001
                        start_dt = datetime.now(UTC) - timedelta(days=overlap_days)
                else:
                    start_dt = parse_yyyy_mm_dd(start_date)

                end_dt = datetime.now(UTC)
                # Oura will reject very large time windows for time-series endpoints
                # (e.g., requesting many years at once), so we fetch in chunks.
                chunk_days = 30
                chunk = timedelta(days=chunk_days)
                print(f"- {name}: {dt_to_iso_z(start_dt)} -> {dt_to_iso_z(end_dt)} ({chunk_days}-day chunks)")
                cur = start_dt
                while cur < end_dt:
                    chunk_end = min(end_dt, cur + chunk)
                    params = {"start_datetime": dt_to_iso_z(cur), "end_datetime": dt_to_iso_z(chunk_end)}
                    items = _oura_fetch_all(sess, access_token=access_token, path=path, params=params)

                    for item in items:
                        # HR samples have timestamps; use them as stable ids.
                        ts = item.get("timestamp") or item.get("time") or item.get("datetime")
                        stable = json.dumps(item, sort_keys=True, separators=(",", ":"), ensure_ascii=True)
                        record_id = str(item.get("id") or ts or sha256_hex(stable))
                        op = db.upsert_record(
                            provider=OURA_PROVIDER,
                            resource=name,
                            record_id=record_id,
                            payload=item,
                            start_time=ts,
                            end_time=None,
                            source_updated_at=ts,
                        )
                        run.add_upsert(op)

                    # Advance window. We don't add an overlap here; record ids dedupe anyway.
                    cur = chunk_end

                db.set_sync_state(provider=OURA_PROVIDER, resource=name, watermark=dt_to_iso_z(end_dt))

    print("Oura sync complete.")

File: src/providers/oura.js
import crypto from 'node:crypto';
import {
  basicAuthHeader,
  dtToIsoZ,
  oauthListenForCode,
  openInBrowser,
  requestJson,
  sha256Hex,
  utcNowIso,
} from '../util.js';

const OURA_BASE = 'https://api.ouraring.com';
const OURA_DEFAULT_AUTHORIZE = 'https://moi.ouraring.com/oauth/v2/ext/oauth-authorize';
const OURA_DEFAULT_TOKEN = 'https://moi.ouraring.com/oauth/v2/ext/oauth-token';

const DATE_WINDOW_RESOURCES = {
  daily_activity: '/v2/usercollection/daily_activity',
  daily_sleep: '/v2/usercollection/daily_sleep',
  daily_readiness: '/v2/usercollection/daily_readiness',
  sleep: '/v2/usercollection/sleep',
  workout: '/v2/usercollection/workout',
};

function randomState() {
  return crypto.randomBytes(16).toString('hex');
}

function parseRedirectConfig(redirectUri) {
  const parsed = new URL(redirectUri);
  return {
    host: parsed.hostname,
    port: parsed.port ? Number.parseInt(parsed.port, 10) : (parsed.protocol === 'https:' ? 443 : 80),
    path: parsed.pathname || '/callback',
    uri: parsed.toString(),
  };
}

function ymdUtc(date) {
  const y = date.getUTCFullYear();
  const m = String(date.getUTCMonth() + 1).padStart(2, '0');
  const d = String(date.getUTCDate()).padStart(2, '0');
  return `${y}-${m}-${d}`;
}

function addDays(date, days) {
  const copy = new Date(date.getTime());
  copy.setUTCDate(copy.getUTCDate() + days);
  return copy;
}

function tokenExpiredSoon(expiresAtIso, skewSeconds = 60) {
  if (!expiresAtIso) {
    return true;
  }
  const expiryMs = Date.parse(expiresAtIso);
  if (Number.isNaN(expiryMs)) {
    return true;
  }
  return expiryMs <= (Date.now() + skewSeconds * 1000);
}

function tokenExtra(raw, endpoint, issuer = null, discoveryUrl = null) {
  const extra = {};
  for (const [key, value] of Object.entries(raw || {})) {
    if (['access_token', 'refresh_token', 'token_type', 'scope', 'expires_in'].includes(key)) {
      continue;
    }
    extra[key] = value;
  }
  extra.token_endpoint = endpoint;
  if (issuer) {
    extra.issuer = issuer;
  }
  if (discoveryUrl) {
    extra.oidc_discovery_url = discoveryUrl;
  }
  return extra;
}

function mergeTokenExtra(existingExtra, freshExtra) {
  return {
    ...(existingExtra || {}),
    ...(freshExtra || {}),
  };
}

function oidcDiscoveryUrls(issuer) {
  if (!issuer || typeof issuer !== 'string') {
    return [];
  }
  const normalized = issuer.replace(/\/$/, '');
  const urls = [`${normalized}/.well-known/openid-configuration`];
  if (normalized.endsWith('/oauth-anonymous')) {
    const parent = normalized.slice(0, -'/oauth-anonymous'.length);
    urls.push(`${parent}/.well-known/openid-configuration`);
  }
  return [...new Set(urls)];
}

async function ouraDiscoverEndpoints(issuer) {
  for (const discoveryUrl of oidcDiscoveryUrls(issuer)) {
    try {
      const doc = await requestJson(discoveryUrl);
      if (doc?.token_endpoint) {
        return {
          tokenEndpoint: String(doc.token_endpoint),
          discoveryUrl,
          issuer,
        };
      }
    } catch {
      // Continue to fallback URLs.
    }
  }
  return null;
}

async function ouraResolveTokenEndpoint(cfg, tokenExtraJson = null, issuerHint = null) {
  if (tokenExtraJson?.token_endpoint) {
    return {
      tokenEndpoint: String(tokenExtraJson.token_endpoint),
      issuer: tokenExtraJson.issuer || issuerHint || null,
      discoveryUrl: tokenExtraJson.oidc_discovery_url || null,
    };
  }

  const issuer = issuerHint || tokenExtraJson?.issuer || null;
  if (issuer) {
    const discovered = await ouraDiscoverEndpoints(String(issuer));
    if (discovered) {
      return discovered;
    }
  }

  return {
    tokenEndpoint: String(cfg.token_url || OURA_DEFAULT_TOKEN),
    issuer,
    discoveryUrl: null,
  };
}

function oauthHeaders(clientId, clientSecret) {
  return {
    Authorization: basicAuthHeader(clientId, clientSecret),
    'Content-Type': 'application/x-www-form-urlencoded',
  };
}

function chooseArray(payload) {
  if (Array.isArray(payload?.data)) {
    return payload.data;
  }
  if (Array.isArray(payload?.result)) {
    return payload.result;
  }
  if (Array.isArray(payload?.items)) {
    return payload.items;
  }
  return [];
}

function dateWindowRecordId(item) {
  return String(item?.id || item?.day || item?.timestamp || sha256Hex(JSON.stringify(item)));
}

function dateWindowStart(item) {
  return item?.day || item?.start_datetime || item?.timestamp || item?.start_time || null;
}

function dateWindowEnd(item) {
  return item?.end_datetime || item?.end_time || null;
}

function dateWindowUpdated(item) {
  return item?.updated_at || item?.modified_at || item?.timestamp || null;
}

function heartrateRecordId(item) {
  return String(item?.id || item?.timestamp || item?.time || item?.datetime || sha256Hex(JSON.stringify(item)));
}

function heartrateTimestamp(item) {
  return item?.timestamp || item?.time || item?.datetime || null;
}

async function ouraFetchAll(pathname, accessToken, params) {
  const out = [];
  let nextToken = null;

  do {
    const requestParams = { ...params };
    if (nextToken) {
      requestParams.next_token = nextToken;
    }

    const payload = await requestJson(`${OURA_BASE}${pathname}`, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
      params: requestParams,
    });

    out.push(...chooseArray(payload));
    nextToken = payload?.next_token || null;
  } while (nextToken);

  return out;
}

async function ouraRefreshIfNeeded(db, cfg) {
  const token = db.getOAuthToken('oura');
  if (!token) {
    throw new Error('Oura token not found. Run `health-sync auth oura`.');
  }
  if (!tokenExpiredSoon(token.expiresAt)) {
    return token.accessToken;
  }

  if (!cfg.client_id || !cfg.client_secret) {
    throw new Error('Oura token expired and missing [oura].client_id/client_secret for refresh.');
  }
  if (!token.refreshToken) {
    throw new Error('Oura refresh_token is missing. Re-run `health-sync auth oura`.');
  }

  const resolved = await ouraResolveTokenEndpoint(cfg, token.extra || null, null);

  let refreshed;
  try {
    refreshed = await requestJson(resolved.tokenEndpoint, {
      method: 'POST',
      headers: oauthHeaders(cfg.client_id, cfg.client_secret),
      data: {
        grant_type: 'refresh_token',
        refresh_token: token.refreshToken,
      },
    });
  } catch (err) {
    const detail = err?.body?.error || err?.body?.error_description || err?.message || String(err);
    if (detail.includes('invalid_grant') || detail.includes('invalid_request')) {
      throw new Error(`Oura refresh failed (${detail}). Re-run \`health-sync auth oura\`.`);
    }
    throw err;
  }

  if (!refreshed?.access_token) {
    throw new Error(`Oura refresh response is missing access_token: ${JSON.stringify(refreshed)}`);
  }

  const expiresIn = Number.parseInt(String(refreshed.expires_in ?? 0), 10) || 0;
  const expiresAt = expiresIn > 0 ? dtToIsoZ(new Date(Date.now() + expiresIn * 1000)) : null;

  const newExtra = tokenExtra(refreshed, resolved.tokenEndpoint, resolved.issuer, resolved.discoveryUrl);
  db.setOAuthToken('oura', {
    accessToken: String(refreshed.access_token),
    refreshToken: refreshed.refresh_token || token.refreshToken,
    tokenType: refreshed.token_type || token.tokenType || 'Bearer',
    scope: refreshed.scope || token.scope,
    expiresAt,
    extra: mergeTokenExtra(token.extra, newExtra),
  });

  return String(refreshed.access_token);
}

async function ouraAuth(db, config, helpers, options = {}) {
  const cfg = helpers.configFor('oura');
  const clientId = helpers.requireStr('oura', 'client_id', 'Missing [oura].client_id');
  const clientSecret = helpers.requireStr('oura', 'client_secret', 'Missing [oura].client_secret');
  const redirectUri = helpers.requireStr('oura', 'redirect_uri', 'Missing [oura].redirect_uri');

  const redirect = parseRedirectConfig(redirectUri);
  const listenHost = options.listenHost || redirect.host || '127.0.0.1';
  const listenPort = Number.isFinite(options.listenPort) && Number(options.listenPort) > 0
    ? Number(options.listenPort)
    : redirect.port;

  const state = randomState();

  const authUrl = new URL(cfg.authorize_url || OURA_DEFAULT_AUTHORIZE);
  authUrl.searchParams.set('response_type', 'code');
  authUrl.searchParams.set('client_id', clientId);
  authUrl.searchParams.set('redirect_uri', redirect.uri);
  authUrl.searchParams.set('scope', String(cfg.scopes || ''));
  authUrl.searchParams.set('state', state);

  openInBrowser(authUrl.toString());
  console.log(`Open this URL to authorize Oura: ${authUrl.toString()}`);

  const callback = await oauthListenForCode({
    listenHost,
    listenPort,
    callbackPath: redirect.path,
    timeoutSeconds: 300,
    allowManualCodeEntry: Boolean(options.allowManualCodeEntry),
    onStatus: (line) => console.log(line),
  });

  if (callback.error) {
    throw new Error(`Oura OAuth error: ${callback.error}`);
  }
  if (!callback.code) {
    throw new Error('Oura OAuth did not return an authorization code');
  }
  if (callback.state && callback.state !== state) {
    throw new Error('Oura OAuth state mismatch');
  }

  const resolved = await ouraResolveTokenEndpoint(cfg, null, callback.issuer || null);

  const tokenPayload = await requestJson(resolved.tokenEndpoint, {
    method: 'POST',
    headers: oauthHeaders(clientId, clientSecret),
    data: {
      grant_type: 'authorization_code',
      code: callback.code,
      redirect_uri: redirect.uri,
    },
  });

  if (!tokenPayload?.access_token) {
    throw new Error(`Oura token response is missing access_token: ${JSON.stringify(tokenPayload)}`);
  }

  const expiresIn = Number.parseInt(String(tokenPayload.expires_in ?? 0), 10) || 0;
  const expiresAt = expiresIn > 0 ? dtToIsoZ(new Date(Date.now() + expiresIn * 1000)) : null;

  db.setOAuthToken('oura', {
    accessToken: String(tokenPayload.access_token),
    refreshToken: tokenPayload.refresh_token || null,
    tokenType: tokenPayload.token_type || 'Bearer',
    scope: tokenPayload.scope || cfg.scopes,
    expiresAt,
    extra: tokenExtra(tokenPayload, resolved.tokenEndpoint, resolved.issuer, resolved.discoveryUrl),
  });

  console.log('Oura authorization succeeded.');
}

function startDateForResource(db, cfg, resource) {
  const overlapDays = Math.max(0, Number.parseInt(String(cfg.overlap_days ?? 7), 10) || 7);
  const state = db.getSyncState('oura', resource);
  if (!state?.watermark) {
    return String(cfg.start_date || '2010-01-01');
  }

  const wmDate = new Date(state.watermark);
  if (Number.isNaN(wmDate.getTime())) {
    return String(cfg.start_date || '2010-01-01');
  }

  const dateOnly = new Date(Date.UTC(
    wmDate.getUTCFullYear(),
    wmDate.getUTCMonth(),
    wmDate.getUTCDate(),
    0,
    0,
    0,
  ));
  dateOnly.setUTCDate(dateOnly.getUTCDate() - overlapDays);
  return ymdUtc(dateOnly);
}

async function syncPersonalInfo(db, token) {
  await db.syncRun('oura', 'personal_info', async () => {
    await db.transaction(async () => {
      const payload = await requestJson(`${OURA_BASE}/v2/usercollection/personal_info`, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });
      db.upsertRecord({
        provider: 'oura',
        resource: 'personal_info',
        recordId: 'me',
        startTime: null,
        endTime: null,
        sourceUpdatedAt: null,
        payload,
      });

      db.setSyncState('oura', 'personal_info', {
        watermark: utcNowIso(),
      });
    });
  });
}

async function syncDateWindowResource(db, token, cfg, resource, endpoint) {
  await db.syncRun('oura', resource, async () => {
    await db.transaction(async () => {
      const startDate = startDateForResource(db, cfg, resource);
      let endDate = ymdUtc(new Date());
      if (resource === 'sleep') {
        endDate = ymdUtc(addDays(new Date(), 1));
      }

      const entries = await ouraFetchAll(endpoint, token, {
        start_date: startDate,
        end_date: endDate,
      });

      for (const item of entries) {
        db.upsertRecord({
          provider: 'oura',
          resource,
          recordId: dateWindowRecordId(item),
          startTime: dateWindowStart(item),
          endTime: dateWindowEnd(item),
          sourceUpdatedAt: dateWindowUpdated(item),
          payload: item,
        });
      }

      db.setSyncState('oura', resource, {
        watermark: utcNowIso(),
      });
    });
  });
}

async function syncHeartrate(db, token, cfg) {
  await db.syncRun('oura', 'heartrate', async () => {
    await db.transaction(async () => {
      const now = new Date();
      const chunkDays = 30;

      const maxStart = db.getMaxRecordStartTime('oura', 'heartrate');
      let cursor = maxStart ? new Date(maxStart) : new Date(`${cfg.start_date || '2010-01-01'}T00:00:00Z`);
      if (Number.isNaN(cursor.getTime())) {
        cursor = new Date('2010-01-01T00:00:00Z');
      }
      cursor = addDays(cursor, -1);

      while (cursor < now) {
        const chunkEnd = new Date(cursor.getTime());
        chunkEnd.setUTCDate(chunkEnd.getUTCDate() + chunkDays);
        if (chunkEnd > now) {
          chunkEnd.setTime(now.getTime());
        }

        const entries = await ouraFetchAll('/v2/usercollection/heartrate', token, {
          start_datetime: dtToIsoZ(cursor),
          end_datetime: dtToIsoZ(chunkEnd),
        });

        for (const item of entries) {
          const ts = heartrateTimestamp(item);
          db.upsertRecord({
            provider: 'oura',
            resource: 'heartrate',
            recordId: heartrateRecordId(item),
            startTime: ts,
            endTime: null,
            sourceUpdatedAt: ts,
            payload: item,
          });
        }

        cursor = addDays(chunkEnd, 0);
        cursor.setUTCSeconds(cursor.getUTCSeconds() + 1);
      }

      db.setSyncState('oura', 'heartrate', {
        watermark: utcNowIso(),
      });
    });
  });
}

async function ouraSync(db, config, helpers) {
  const cfg = helpers.configFor('oura');
  const token = await ouraRefreshIfNeeded(db, cfg);

  await syncPersonalInfo(db, token);
  for (const [resource, endpoint] of Object.entries(DATE_WINDOW_RESOURCES)) {
    await syncDateWindowResource(db, token, cfg, resource, endpoint);
  }
  await syncHeartrate(db, token, cfg);
}

const ouraProvider = {
  id: 'oura',
  source: 'builtin',
  description: 'Oura daily data, sleep/workout collections, heartrate, and personal info',
  supportsAuth: true,
  auth: ouraAuth,
  sync: ouraSync,
};

export default ouraProvider;
